client-core/src/service/key_service.rs

```diff
@@ -56,7 +92,152 @@ where
+        let seed = Seed::new(&mnemonic, "");
+        self.storage
+            .set_secure(KEYSPACE_HD, name, seed.as_bytes().into(), passphrase)?;
+        debug!("hdwallet write seed success");
+        Ok(())
+    }
+
+    /// auto-matically generate staking, transfer addresses
+    /// with just one api call
+    #[allow(dead_code)]
+    pub fn auto_restore(
+        &self,
+        mnemonic: &str,
+        name: &str,
+        passphrase: &SecUtf8,
+        count: i32,
+    ) -> Result<()> {
+        self.generate_seed(mnemonic, name, passphrase)
+            .expect("auto restore");
+        let cointype = get_bip44_coin_type();
+        println!("coin type={}", cointype);
+        for index in 0..count {
+            let seed_bytes = self.storage.get_secure(KEYSPACE_HD, name, passphrase)?;
+            for account in 0..2 {
+                let extended = ExtendedPrivKey::derive(
+                    &seed_bytes.clone().unwrap()[..],
+                    format!("m/44'/{}'/{}'/0/{}", cointype, account, index).as_str(),
+                )
+                .unwrap();
+                let secret_key_bytes = extended.secret();
+
+                let private_key = PrivateKey::deserialize_from(&secret_key_bytes).unwrap();
+                let public_key = PublicKey::from(&private_key);
+
+                self.storage.set_secure(
+                    KEYSPACE,
+                    public_key.serialize(),
+                    private_key.serialize(),
+                    passphrase,
+                )?;
+            }
+        }
+        Ok(())
+    }
+
+    /// read value from db, if it's None, there value doesn't exist
+    pub fn read_value(&self, passphrase: &SecUtf8, key: &[u8]) -> Option<Vec<u8>> {
+        if let Ok(connected) = self.storage.get_secure(KEYSPACE_HD, key, passphrase) {
+            if let Some(value) = connected {
+                return Some(value.clone());
+            }
+        }
+        None
+    }
+
+    /// read number, if value doesn't exist, it returns default value
+    pub fn read_number(&self, passphrase: &SecUtf8, key: &[u8], default: u32) -> u32 {
+        if let Ok(connected) = self.storage.get_secure(KEYSPACE_HD, key, passphrase) {
+            if let Some(value) = connected {
+                return std::str::from_utf8(&value[..])
+                    .unwrap()
+                    .parse::<u32>()
+                    .unwrap();
+            }
+        }
+        default
+    }
+
+    /// write number to store, write number as string
+    /// writes hdwallet index, after making a new entry, index increases by 1
+    /// so address is generated in deterministic way.
+    pub fn write_number(&self, passphrase: &SecUtf8, key: &[u8], value: u32) {
+        let a = value.to_string();
+        let b = a.as_bytes();
+        self.storage
+            .set_secure(KEYSPACE_HD, key, b.to_vec(), passphrase)
+            .unwrap();
```

Comment:
Can you handle error instead of `unwrap`ing?

Reply:
ok